ambry-replication/src/main/java/com.github.ambry.replication/ReplicationEngine.java

```diff
@@ -0,0 +1,502 @@
+import com.github.ambry.clustermap.PartitionId;
+import com.github.ambry.commons.ResponseHandler;
+import com.github.ambry.config.ClusterMapConfig;
+import com.github.ambry.config.ReplicationConfig;
+import com.github.ambry.network.ConnectionPool;
+import com.github.ambry.notification.NotificationSystem;
+import com.github.ambry.store.FindToken;
+import com.github.ambry.store.FindTokenFactory;
+import com.github.ambry.store.StoreKeyConverter;
+import com.github.ambry.store.StoreKeyConverterFactory;
+import com.github.ambry.store.StoreKeyFactory;
+import com.github.ambry.store.Transformer;
+import com.github.ambry.utils.CrcInputStream;
+import com.github.ambry.utils.SystemTime;
+import com.github.ambry.utils.Utils;
+import java.io.DataInputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Abstract class to handle replication for given partitions. Responsible for the following
+ * 1. Initialize class members.
+ * 2. Create replica threads and distribute partitions amongst the threads
+ * 3. Set up replica token persistor used to recover from shutdown/crash
+ * 4. Initialize and shutdown all the components required to perform replication
+ */
+public abstract class ReplicationEngine {
+
+  private final ReplicationConfig replicationConfig;
+  private final ClusterMap clusterMap;
+  private final ScheduledExecutorService scheduler;
+  private final AtomicInteger correlationIdGenerator;
+  private final ConnectionPool connectionPool;
+  private final NotificationSystem notification;
+  private final Map<String, DataNodeRemoteReplicaInfos> dataNodeRemoteReplicaInfosPerDC;
+  private final StoreKeyFactory storeKeyFactory;
+  private final List<String> sslEnabledDatacenters;
+  private final Map<String, List<ReplicaThread>> replicaThreadPools;
+  private final StoreKeyConverterFactory storeKeyConverterFactory;
+  private final String transformerClassName;
+
+  protected final DataNodeId dataNodeId;
+  protected final MetricRegistry metricRegistry;
+  protected final ReplicationMetrics replicationMetrics;
+  protected final FindTokenFactory factory;
+  protected final Logger logger = LoggerFactory.getLogger(getClass());
+  protected final Map<PartitionId, PartitionInfo> partitionsToReplicate;
+  protected final Map<String, List<PartitionInfo>> partitionGroupedByMountPath;
+  protected final Map<String, Integer> numberOfReplicaThreads;
+  protected ReplicaTokenPersistor persistor = null;
+
+  static final short Replication_Delay_Multiplier = 5;
```

Comment:
Minor: if this is only used in ReplicationManager, we can move it there and make private.

Reply:
Protected.